activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -4,6 +4,7 @@ module QueryCache
       class << self
         def included(base) #:nodoc:
           dirties_query_cache base, :insert, :update, :delete, :rollback_to_savepoint, :rollback_db_transaction
+          base.set_callback :checkin, :after, :disable_query_cache!
```

Comment:
We should probably clear the query cache before the connection is checked in, rather than afterwards.

If the callback chain halts before this one can run - if there is another after callback configured, and it raises, for example - the connection would end up in the pool with a dirty cache.